docs/sm.8
.nh
.TH sm 8

.SH NAME
.PP
sm - AWS Secrets Manager CLI Tool


.SH SYNOPSIS
.PP
sm


.SH COMMAND TREE
.RS
.IP \(bu 2
get, view
\[la]#get-view\[ra]
.IP \(bu 2
edit, e
\[la]#edit-e\[ra]
.IP \(bu 2
create, c
\[la]#create-c\[ra]
.IP \(bu 2
put
\[la]#put\[ra]
.IP \(bu 2
delete, del
\[la]#delete-del\[ra]
.IP \(bu 2
list
\[la]#list\[ra]
.IP \(bu 2
describe
\[la]#describe\[ra]

.RE

.PP
\fBUsage\fP:

.PP
.RS

.nf
sm [GLOBAL OPTIONS] command [COMMAND OPTIONS] [ARGUMENTS...]

.fi
.RE


.SH COMMANDS
.SH get, view
.PP
select from list or pass in specific secret

.PP
\fB--binary, -b\fP: get the SecretBinary value

.PP
\fB--secret-id, -s\fP="": Specific Secret to view, will bypass select/search

.SH edit, e
.PP
interactive edit of a secret String Value

.PP
\fB--binary, -b\fP: get the SecretBinary value

.PP
\fB--secret-id, -s\fP="": Specific Secret to edit, will bypass select/search

.SH create, c
.PP
create new secret in Secrets Manager

.PP
\fB--binary, -b\fP: get the SecretBinary value

.PP
\fB--description, -d\fP="": Additional description text.

.PP
\fB--interactive, -i\fP: Open interactive editor to create secret value. If no 'value' is provided, an editor will be opened by default.

.PP
\fB--secret-id, -s\fP="": Secret name

.PP
\fB--tags, -t\fP="": key=value tags (CSV list)

.PP
\fB--value, -v\fP="": Secret Value. Will store as a string, unless binary flag is set.

.SH put
.PP
non-interactive update to a specific secret

.PP
.RS

.nf
Stores a new encrypted secret value in the specified secret. To do this, the 
operation creates a new version and attaches it to the secret. The version 
can contain a new SecretString value or a new SecretBinary value.

This will put the value to AWSCURRENT and retain one previous version 
with AWSPREVIOUS.

.fi
.RE

.PP
\fB--binary, -b\fP: get the SecretBinary value

.PP
\fB--interactive, -i\fP: Override and open interactive editor to verify and modify the new secret value.

.PP
\fB--secret-id, -s\fP="": Secret name

.PP
\fB--value, -v\fP="": Secret Value. Will store as a string, unless binary flag is set.

.SH delete, del
.PP
delete a specific secret

.PP
\fB--force, -f\fP: Bypass recovery window (30 days) and immediately delete Secret.

.PP
\fB--secret-id, -s\fP="": Specific Secret to delete

.SH list
.PP
display table of all secrets with meta data

.SH describe
.PP
print description of secret to \fB\fCSTDOUT\fR

.PP
\fB--secret-id, -s\fP="": Specific Secret to describe, will bypass select/search



CHANGELOG.md
<a name="unreleased"></a>
## [Unreleased]


<a name="v0.1.2"></a>
## [v0.1.2] - 2022-05-23
### Chore
- **docs:** updating docs for version v0.1.2
  
  ### Pull Requests
- chore(deps): update goreleaser/goreleaser-action action to v3 ([#14](https://github.com/clok/sm/issues/14))
  - chore(deps): update docker/login-action action to v2 ([#13](https://github.com/clok/sm/issues/13))
  - fix(deps): update all non-major dependencies ([#12](https://github.com/clok/sm/issues/12))
  - chore(deps): update all non-major dependencies ([#8](https://github.com/clok/sm/issues/8))
  - chore(deps): update actions/setup-go action to v3 ([#10](https://github.com/clok/sm/issues/10))
  - chore(deps): update actions/checkout action to v3 ([#11](https://github.com/clok/sm/issues/11))
  - chore(deps): update golangci/golangci-lint-action action to v3 ([#9](https://github.com/clok/sm/issues/9))
  
  
<a name="v0.1.1"></a>
## [v0.1.1] - 2022-01-21
### Chore
- **release:** v0.1.1
  
  
<a name="v0.1.0-8-ge10cff7"></a>
## [v0.1.0-8-ge10cff7] - 2022-01-21
### Bug Fixes
- **deps:** update all non-major dependencies
  
  ### Chore
- update readme
  - **release:** v0.1.0-8-ge10cff7
  
  ### Pull Requests
- fix(deps): update module github.com/aws/aws-sdk-go to v1.42.39 ([#7](https://github.com/clok/sm/issues/7))
  - fix(deps): update all non-major dependencies ([#6](https://github.com/clok/sm/issues/6))
  - chore(deps): update all non-major dependencies docker tags ([#5](https://github.com/clok/sm/issues/5))
  - chore(deps): update all non-major dependencies docker tags ([#3](https://github.com/clok/sm/issues/3))
  - fix(deps): update all non-major dependencies ([#2](https://github.com/clok/sm/issues/2))
  - Merge pull request [#1](https://github.com/clok/sm/issues/1) from clok/renovate/all-minor-patch
  
  
<a name="v0.1.0"></a>
## v0.1.0 - 2021-06-17
### Chore
- clean up main and reorg the commands
  - **docs:** updating docs for version v0.1.0
  - **docs:** updated README
  - **release:** v0.1.0
  
  ### Ci
- add all automation configuration
  
  ### Features
- **sm:** add initial actions for AWS secrets manager
  
  
[Unreleased]: https://github.com/clok/sm/compare/v0.1.2...HEAD
[v0.1.2]: https://github.com/clok/sm/compare/v0.1.1...v0.1.2
[v0.1.1]: https://github.com/clok/sm/compare/v0.1.0-8-ge10cff7...v0.1.1
[v0.1.0-8-ge10cff7]: https://github.com/clok/sm/compare/v0.1.0...v0.1.0-8-ge10cff7


cmd/sm.go
package cmd

import (
	"encoding/json"
	"fmt"
	"os"
	"sort"
	"strings"

	"github.com/AlecAivazis/survey/v2"
	"github.com/TylerBrock/colorjson"
	"github.com/a8m/djson"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/clok/sm/aws"
	"github.com/jedib0t/go-pretty/v6/table"
	"github.com/urfave/cli/v2"
)

// truncateString limits the length of a string while also appending an ellipses.
func truncateString(str string, num int) string {
	short := str
	if len(str) > num {
		if num > 3 {
			num -= 3
		}
		short = str[0:num] + "..."
	}
	return short
}

// selectSecretNameFromList is a helper method to either bypass and return the
// `secretName` passed in via CLI flag OR retrieve a list of all secrets to allow
// for a search select by the User.
func selectSecretNameFromList(c *cli.Context) (string, error) {
	secretName := c.String("secret-id")
	if secretName == "" {
		secrets, err := sm.ListSecrets()
		if err != nil {
			PrintWarn("Error retrieving list of secrets.")
			return "", err
		}

		secretNames := make([]string, 0, len(secrets))
		for _, secret := range secrets {
			secretNames = append(secretNames, aws.StringValue(secret.Name))
		}
		sort.Strings(secretNames)

		p := &survey.Select{
			Message: "Choose a Secret to view:",
			Options: secretNames,
			Default: secretNames[0],
		}
		err = survey.AskOne(p, &secretName)
		if err != nil {
			return "", err
		}

		PrintInfo(fmt.Sprintf("Retrieving: %s", secretName))
	}
	return secretName, nil
}

// promptForEdit is a helper method providing an editor interface.
func promptForEdit(secretName string, s []byte) ([]byte, error) {
	ed := ""
	prompt := &survey.Editor{
		Message:       fmt.Sprintf("Open editor to modify '%s'?", secretName),
		FileName:      "*.json",
		Default:       string(s),
		HideDefault:   true,
		AppendDefault: true,
	}
	err := survey.AskOne(prompt, &ed, nil)
	if err != nil {
		return nil, err
	}

	return []byte(ed), nil
}

// validateAndUpdateSecretValue will take the original buffer and interactively open an
// editor to allow for updates. It will check if the original and updated buffer match, if
// so it will exit gracefully. It will also verify valid JSON and prompt if an invalid input
// is provided.
func validateAndUpdateSecretValue(secretName string, orig []byte, updateTmp []byte) ([]byte, error) {
	done := false
	for !done {
		_, err := djson.Decode(updateTmp)
		if err != nil {
			PrintWarn("invalid JSON submitted.")

			ed := false
			p1 := &survey.Confirm{
				Message: "Open to edit?",
			}
			err = survey.AskOne(p1, &ed)
			if err != nil {
				return nil, err
			}
			if ed {
				updateTmp, err = promptForEdit(secretName, updateTmp)
				if err != nil {
					return nil, cli.Exit(err, 2)
				}
				if string(orig) == strings.TrimSuffix(string(updateTmp), "\n") {
					PrintInfo("Updated value matches original. Exiting.")
					return nil, cli.Exit("", 0)
				}
			} else {
				submit := false
				p2 := &survey.Confirm{
					Message: "Continue with Submit?",
				}
				err = survey.AskOne(p2, &submit)
				if err != nil {
					return nil, err
				}
				if !submit {
					PrintWarn("Exiting without submit.")
					return nil, cli.Exit("", 0)
				}
				PrintInfo("Continuing with submit.")
				done = true
			}
		} else {
			PrintInfo("JSON validated.")
			done = true
		}
	}
	return updateTmp, nil
}

// ListSecrets CLI command to list all Secrets.
func ListSecrets(c *cli.Context) error {
	secrets, err := sm.ListSecrets()
	if err != nil {
		return cli.Exit(err, 2)
	}

	t := table.NewWriter()
	t.SetOutputMirror(os.Stdout)
	t.SetStyle(table.StyleLight)
	t.AppendHeader(table.Row{"Name", "Updated", "Accessed", "Description"})
	t.SetColumnConfigs([]table.ColumnConfig{
		{Name: "Name", WidthMax: 120},
		{Name: "Updated", WidthMax: 10},
		{Name: "Accessed", WidthMax: 10},
		{
			Name:     "Description",
			WidthMax: 40,
		},
	})
	t.SortBy([]table.SortBy{
		{Name: "Name", Mode: table.Asc},
	})

	for _, secret := range secrets {
		lastdt := aws.TimeValue(secret.LastAccessedDate)
		updateddt := aws.TimeValue(secret.LastChangedDate)
		t.AppendRow([]interface{}{
			aws.StringValue(secret.Name),
			fmt.Sprintf("%d-%02d-%02d", updateddt.Year(), updateddt.Month(), updateddt.Day()),
			fmt.Sprintf("%d-%02d-%02d", lastdt.Year(), lastdt.Month(), lastdt.Day()),
			truncateString(aws.StringValue(secret.Description), 40),
		})
	}

	t.Render()

	return nil
}

// ViewSecret CLI command to view/get a Secret.
func ViewSecret(c *cli.Context) error {
	secretName, err := selectSecretNameFromList(c)
	if err != nil {
		return cli.Exit(err, 2)
	}

	secret, err := sm.GetSecret(secretName)
	if err != nil {
		return cli.Exit(err, 2)
	}

	if c.Bool("binary") {
		fmt.Println(string(secret.SecretBinary))
	} else {
		result, err := djson.Decode([]byte(aws.StringValue(secret.SecretString)))
		if err != nil {
			PrintWarn("stored string value is not valid JSON.")
			fmt.Println(aws.StringValue(secret.SecretString))
		} else {
			f := colorjson.NewFormatter()
			f.Indent = 4

			s, _ := f.Marshal(result)
			fmt.Println(string(s))
		}
	}

	return nil
}

// DescribeSecret CLI command to describe a Secret.
func DescribeSecret(c *cli.Context) error {
	secretName, err := selectSecretNameFromList(c)
	if err != nil {
		return cli.Exit(err, 2)
	}

	secret, err := sm.DescribeSecret(secretName)
	if err != nil {
		return cli.Exit(err, 2)
	}

	fmt.Println(secret.String())

	return nil
}

// EditSecret CLI command to edit a Secret.
func EditSecret(c *cli.Context) error {
	secretName, err := selectSecretNameFromList(c)
	if err != nil {
		return cli.Exit(err, 2)
	}

	secret, err := sm.GetSecret(secretName)
	if err != nil {
		return cli.Exit(err, 2)
	}

	var s []byte
	if c.Bool("binary") {
		s = secret.SecretBinary
	} else {
		result, err := djson.Decode([]byte(aws.StringValue(secret.SecretString)))
		if err != nil {
			PrintWarn("stored string value is not valid JSON.")
			s = []byte(aws.StringValue(secret.SecretString))
		} else {
			s, err = json.MarshalIndent(result, "", "    ")
			if err != nil {
				return cli.Exit(err, 2)
			}
		}
	}

	var up []byte
	up, err = promptForEdit(secretName, s)
	if err != nil {
		return cli.Exit(err, 2)
	}
	if string(s) == strings.TrimSuffix(string(up), "\n") {
		PrintInfo("Updated value matches original. Exiting.")
		return nil
	}

	var final []byte
	final, err = validateAndUpdateSecretValue(secretName, s, up)
	if err != nil {
		return cli.Exit(err, 2)
	}

	var t string
	if c.Bool("binary") {
		t = "BinarySecret"
		_, err = sm.PutSecretBinary(secretName, final)
	} else {
		t = "StringSecret"
		_, err = sm.PutSecretString(secretName, string(final))
	}

	if err != nil {
		return cli.Exit(err, 2)
	}

	PrintSuccess(fmt.Sprintf("%s %s successfully updated.", secretName, t))

	return nil
}

// CreateSecret CLI command to create a new Secret.
func CreateSecret(c *cli.Context) error {
	secretName := c.String("secret-id")
	exists, err := sm.CheckIfSecretExists(secretName)
	if err != nil {
		return cli.Exit(err, 2)
	}
	if exists {
		PrintWarn(fmt.Sprintf("'%s' already exists. Please use a different name.", secretName))
		return nil
	}

	interactive := c.Bool("interactive")
	var value []byte
	if c.String("value") == "" {
		// Assume interactive mode
		interactive = true
		value = []byte("{}")
	} else {
		value = []byte(c.String("value"))
	}

	var s []byte
	if interactive {
		result, err := djson.Decode(value)
		if err != nil {
			PrintWarn("value is not valid JSON.")
			s = value
		} else {
			s, err = json.MarshalIndent(result, "", "    ")
			if err != nil {
				return cli.Exit(err, 2)
			}
		}

		var up []byte
		up, err = promptForEdit(secretName, s)
		if err != nil {
			return cli.Exit(err, 2)
		}
		s = up
	} else {
		s = value
	}

	var t string
	if c.Bool("binary") {
		t = "BinarySecret"
		_, err = sm.CreateSecretBinary(secretName, s, c.String("description"), c.String("tags"))
	} else {
		t = "StringSecret"
		_, err = sm.CreateSecretString(secretName, string(s), c.String("description"), c.String("tags"))
	}

	if err != nil {
		return cli.Exit(err, 2)
	}

	PrintSuccess(fmt.Sprintf("%s %s successfully created.", secretName, t))

	return nil
}

// PutSecret CLI command to apply a delta to a Secret.
func PutSecret(c *cli.Context) error {
	secretName := c.String("secret-id")
	exists, err := sm.CheckIfSecretExists(secretName)
	if err != nil {
		return cli.Exit(err, 2)
	}
	if !exists {
		PrintWarn(fmt.Sprintf("'%s' does not exists. Please create the secret first.", secretName))
		return nil
	}

	interactive := c.Bool("interactive")
	var value []byte
	if c.String("value") == "" {
		// Assume interactive mode
		interactive = true

		secret, err := sm.GetSecret(secretName)
		if err != nil {
			return cli.Exit(err, 2)
		}

		if c.Bool("binary") {
			value = secret.SecretBinary
		} else {
			result, err := djson.Decode([]byte(aws.StringValue(secret.SecretString)))
			if err != nil {
				PrintWarn("stored string value is not valid JSON.")
				value = []byte(aws.StringValue(secret.SecretString))
			} else {
				value, err = json.MarshalIndent(result, "", "    ")
				if err != nil {
					return cli.Exit(err, 2)
				}
			}
		}
	} else {
		value = []byte(c.String("value"))
	}

	var final []byte
	if interactive {
		var up []byte
		up, err = promptForEdit(secretName, value)
		if err != nil {
			return cli.Exit(err, 2)
		}
		if string(value) == strings.TrimSuffix(string(up), "\n") {
			PrintInfo("Updated value matches original. Exiting.")
			return nil
		}

		final, err = validateAndUpdateSecretValue(secretName, value, up)
		if err != nil {
			return cli.Exit(err, 2)
		}
	} else {
		final = value
	}

	var t string
	if c.Bool("binary") {
		t = "BinarySecret"
		_, err = sm.PutSecretBinary(secretName, final)
	} else {
		t = "StringSecret"
		_, err = sm.PutSecretString(secretName, string(final))
	}

	if err != nil {
		return cli.Exit(err, 2)
	}

	PrintSuccess(fmt.Sprintf("%s %s successfully put new version.", secretName, t))

	return nil
}

// DeleteSecret CLI command that will delete a Secret.
func DeleteSecret(c *cli.Context) error {
	secretName := c.String("secret-id")
	exists, err := sm.CheckIfSecretExists(secretName)
	if err != nil {
		return cli.Exit(err, 2)
	}
	if !exists {
		PrintWarn(fmt.Sprintf("'%s' was not found.", secretName))
		return nil
	}

	del := false
	p1 := &survey.Confirm{
		Message: fmt.Sprintf("Are you sure you want to permanentaly delete '%s'?", secretName),
	}
	err = survey.AskOne(p1, &del)
	if err != nil {
		return cli.Exit(err, 2)
	}

	if !del {
		PrintInfo("Exiting without delete.")
		return nil
	}

	force := c.Bool("force")
	_, err = sm.DeleteSecret(secretName, force)
	if err != nil {
		return cli.Exit(err, 2)
	}

	PrintSuccess(fmt.Sprintf("'%s' deleted. (force: %v)", secretName, force))

	return nil
}
